src/OpenTelemetry/Metrics/MetricReader.cs

```diff
@@ -48,11 +49,30 @@ public AggregationTemporality SupportedAggregationTemporality
             }
         }
 
-        public virtual void Collect()
+        public virtual bool Collect(int timeoutMilliseconds = Timeout.Infinite)
         {
+            var sw = Stopwatch.StartNew();
+
             var collectMetric = this.ParentProvider.GetMetricCollect();
             var metricsCollected = collectMetric();
-            this.OnCollect(metricsCollected);
+
+            if (timeoutMilliseconds == Timeout.Infinite)
+            {
+                this.OnCollect(metricsCollected, Timeout.Infinite);
+            }
+            else
+            {
+                var timeout = timeoutMilliseconds - sw.ElapsedMilliseconds;
+
+                if (timeout <= 0)
+                {
+                    return false;
```

Comment:
if time already exceeded, then we are just dropping the metrics here, right?

Reply:
Yes, this is my thinking:

For flush and collect, if we run out of time, we'll just skip.

For shutdown, if we run out of time, we'll need to make sure the underlying components got the shutdown notification (with an indication that they have 0 timeout), so they can decide what to do.